File: web/src/utils/textFile.ts
// Определение текстовых файлов: mime text/* + исходники языков
// программирования и конфиги по расширению (их mime может быть
// application/octet-stream).
const TEXT_EXTS = new Set([
  // Исходники
  'py', 'js', 'jsx', 'ts', 'tsx', 'vue', 'go', 'rs', 'java', 'kt', 'c', 'h', 'cpp', 'hpp',
  'cs', 'php', 'rb', 'swift', 'dart', 'lua', 'pl', 'r', 'm', 'mm', 'sh', 'bash', 'zsh',
  'ps1', 'bat', 'cmd', 'sql', 'html', 'htm', 'css', 'scss', 'less', 'xml', 'json', 'yml',
  'yaml', 'toml', 'ini', 'conf', 'cfg', 'env', 'properties', 'gradle', 'tf', 'proto',
  'dockerfile', 'makefile', 'cmake', 'nim', 'ex', 'exs', 'erl', 'hs', 'clj', 'scala',
  'groovy', 'vb', 'pas', 'asm', 's', 'fs', 'fsx', 'ml', 'v', 'zig', 'odin', 'cr', 'go2',
  // Тексты и разметка
  'txt', 'md', 'markdown', 'rst', 'adoc', 'tex', 'log', 'csv', 'tsv', 'diff', 'patch',
  'svg', 'graphql', 'prisma',
])

export function isTextFile(mime: string, filename: string): boolean {
  if (mime.startsWith('text/')) return true
  const name = filename.toLowerCase()
  const ext = name.split('.').pop() || ''
  return TEXT_EXTS.has(ext) || name === 'dockerfile' || name === 'makefile'
}

// Язык подсветки (highlight.js) по расширению файла. Автоопределение по
// содержимому ненадёжно (на коротких фрагментах highlight.js ошибается),
// а имя файла даёт точный ответ.
const EXT_LANG: Record<string, string> = {
  go: 'go', py: 'python', js: 'javascript', jsx: 'javascript', ts: 'typescript', tsx: 'typescript',
  vue: 'vue', rs: 'rust', java: 'java', kt: 'kotlin', c: 'c', h: 'c', cpp: 'cpp', hpp: 'cpp',
  cc: 'cpp', cs: 'csharp', php: 'php', rb: 'ruby', swift: 'swift', dart: 'dart', lua: 'lua',
  pl: 'perl', r: 'r', m: 'objectivec', sh: 'bash', bash: 'bash', zsh: 'bash', ps1: 'powershell',
  bat: 'bat', cmd: 'batch', sql: 'sql', html: 'xml', htm: 'xml', xml: 'xml', css: 'css',
  scss: 'scss', less: 'less', json: 'json', yml: 'yaml', yaml: 'yaml', toml: 'ini', ini: 'ini',
  md: 'markdown', markdown: 'markdown', diff: 'diff', gradle: 'gradle', tf: 'hcl',
  proto: 'protobuf', erl: 'erlang', ex: 'elixir', exs: 'elixir', hs: 'haskell', clj: 'clojure',
  scala: 'scala', groovy: 'groovy', vb: 'vbnet', pas: 'delphi', asm: 'nasm', fs: 'fsharp',
  fsx: 'fsharp', tex: 'tex', csv: 'plaintext', log: 'plaintext', txt: 'plaintext',
}

export function langByFilename(filename: string): string | null {
  const name = filename.toLowerCase()
  if (name === 'dockerfile') return 'dockerfile'
  if (name === 'makefile') return 'makefile'
  const ext = name.split('.').pop() || ''
  return EXT_LANG[ext] || null
}

File: mobile/lib/src/utils/text_file.dart
// Определение текстовых файлов — 1:1 с web/src/utils/textFile.ts
library;

const _textExts = {
  'txt', 'md', 'markdown', 'rst', 'adoc', 'tex', 'log', 'csv', 'tsv', 'diff', 'patch',
  'json', 'yaml', 'yml', 'toml', 'ini', 'cfg', 'conf', 'env', 'sh', 'bash', 'zsh',
  'py', 'js', 'ts', 'jsx', 'tsx', 'dart', 'go', 'rs', 'java', 'kt', 'swift', 'c', 'h', 'cpp', 'hpp', 'cs', 'php', 'rb', 'lua', 'sql', 'html', 'css', 'scss', 'less', 'vue', 'svelte', 'xml', 'gradle', 'tf', 'hcl', 'dockerfile', 'makefile',
};

const _extLang = {
  'txt': 'plaintext', 'md': 'markdown', 'markdown': 'markdown', 'json': 'json', 'yaml': 'yaml', 'yml': 'yaml',
  'toml': 'toml', 'ini': 'ini', 'sh': 'bash', 'bash': 'bash', 'zsh': 'bash', 'py': 'python', 'js': 'javascript',
  'ts': 'typescript', 'jsx': 'javascript', 'tsx': 'typescript', 'dart': 'dart', 'go': 'go', 'rs': 'rust',
  'java': 'java', 'kt': 'kotlin', 'swift': 'swift', 'c': 'c', 'h': 'c', 'cpp': 'cpp', 'hpp': 'cpp',
  'cs': 'csharp', 'php': 'php', 'rb': 'ruby', 'lua': 'lua', 'sql': 'sql', 'html': 'xml', 'css': 'css',
  'scss': 'scss', 'less': 'less', 'vue': 'vue', 'svelte': 'svelte', 'xml': 'xml', 'gradle': 'gradle', 'tf': 'hcl',
};

bool isTextFile(String filename) {
  final parts = filename.toLowerCase().split('.');
  if (parts.length < 2) return false;
  final ext = parts.last;
  return _textExts.contains(ext);
}

String extLang(String filename) {
  final ext = filename.toLowerCase().split('.').last;
  return _extLang[ext] ?? 'plaintext';
}
